pkg/handlers/v1/ec2.go

```diff
@@ -69,40 +78,113 @@ type ec2ChangedPropsARN struct {
 
 type ec2Transformer struct{}
 
-func (t ec2Transformer) Create(event awsConfigEvent) (Output, error) {
-	output, err := getBaseOutput(event.ConfigurationItem)
+func (t ec2Transformer) Create(event awsConfigEvent) ([]Output, error) {
+	outputs := []Output{}
+	baseOutput, err := getBaseOutput(event.ConfigurationItem)
 	if err != nil {
-		return Output{}, err
+		return []Output{}, err
 	}
-
 	// if a resource is created for the first time, there is no diff.
 	// just read the configuration
 	var config ec2Configuration
 	if err := json.Unmarshal(event.ConfigurationItem.Configuration, &config); err != nil {
-		return Output{}, err
+		return []Output{}, err
 	}
-	change := extractEC2NetworkInfo(&config)
-	change.ChangeType = added
-	output.Changes = append(output.Changes, change)
-	return output, nil
+
+	hasEni := false
+	for _, ni := range config.NetworkInterfaces {
+		ni := ni
+		hasEni = true
+		output := baseOutput
+		change := Change{}
+		output.ChangeTime = ni.Attachment.AttachTime
+		private, public, dns := extractNetworkInterfaceInfo(&ni)
+		change.PrivateIPAddresses = append(change.PrivateIPAddresses, private...)
+		change.PublicIPAddresses = append(change.PublicIPAddresses, public...)
+		change.Hostnames = append(change.Hostnames, dns...)
+		change.ChangeType = added
+		output.Changes = append(output.Changes, change)
+		outputs = append(outputs, output)
+	}
+	if !hasEni {
+		return []Output{}, nil // Reject EC2s with no ENI
+	}
+
+	return outputs, nil
 }
 
-func (t ec2Transformer) Update(event awsConfigEvent) (Output, error) {
-	output, err := getBaseOutput(event.ConfigurationItem)
+func initializeAttachMapEntry(attachMap map[string]map[string][]string, attachTime string) {
+	attachMap[attachTime] = make(map[string][]string)
+	attachMap[attachTime][private] = make([]string, 0)
+	attachMap[attachTime][public] = make([]string, 0)
+	attachMap[attachTime][dns] = make([]string, 0)
+}
+
+func separateAddedChanges(baseOutput Output, addedChange Change, privateIPMap map[string]string, publicIPMap map[string]string, hostnameMap map[string]string) []Output {
```

Comment:
Nit: I think it would be better to do this at the ENI level. It may require a bit of refactoring to the "symmetricDifference" function, but I think that is more fool proof and easier to read and reason about. Then you won't need such a complicated method that almost reconstructs the ENI.

Reply:
Ashley and I chatted about this, there could be some weird edge cases that complicate my suggested method a little. Ex: The same IP was deleted via one ENI and reattached via another. While what I suggested could still work, it would result in likely a 3 tiered for loop which may be more complicated than what this is doing. So I'm okay with keeping this as is.